Provision an instance of CartoDB on Vagrant

CloudNiner opened the pull request:
Generally follows the install instructions from:
http://cartodb.readthedocs.org/en/latest/install.html

Instructions for use are in README.md

Commit:
Provision an instance of CartoDB on Vagrant

Generally follows the install instructions from:
http://cartodb.readthedocs.org/en/latest/install.html

Instructions for use are in README.md

Commit:
Move windshaft/sqlapi configs to ansible templates

Modify the host parameter of each config so we can have
  these services listen directly on 0.0.0.0 rather
  than having to setup nginx to proxy requests.

Could use this in the future as a springboard for staging|production
  templates or to further customize the config.

If the config changes significantly in future versions of sqlapi
  or windshaft, it will need to be updated here as well.

CloudNiner commented:
This PR does not currently fit the format of our other `ansible-*` repositories. Additional work would need to be done for that. But I figured that wasn't terribly pressing atm since our primary use case is to use this repo as a provisioner for a cartodb dev environment, rather than using it in other ansible roles.

ddohler commented on ansible/roles/cartodb.app/tasks/main.yml:
```diff
@@ -0,0 +1,51 @@
+
+ # TODO: Template if necessary
+ - name: Create app_config.yml
+   command: cp config/app_config.yml.sample config/app_config.yml
```
It might be good to add a `creates=` specification here so that this doesn't clobber `app_config.yml` on consecutive runs; same for the `database.yml` one below.

CloudNiner replied:
Ah, fair enough. Will do.

ddohler commented on ansible/roles/cartodb.postgresql/tasks/main.yml:
```diff
@@ -0,0 +1,55 @@
+   command: createuser publicuser --no-createrole --no-createdb --no-superuser
+   sudo: True
+   sudo_user: postgres
+   # TODO: Come up with better way to handle user already created
```
There's a `postgresql_user` module -- would that work? http://docs.ansible.com/ansible/postgresql_user_module.html

CloudNiner replied:
Sort of. I punted because the helpful `azavea.postgresql-support` role, which installs the dependencies for that module, has packages that conflict with the cartodb postgresql ppa. I figured it was something we could create an issue for and look at later, since the error message this throws (and skips) when the user already exists is very clear in the ansible log and I had more pressing things to try to figure out as part of this first iteration.

ddohler commented on ansible/roles/cartodb.app/tasks/main.yml:
```diff
@@ -0,0 +1,51 @@
+     chdir: "{{ cartodb_dir }}"
+
+ - name: Create a cartodb user
+   command: bundle exec rake cartodb:db:create_user SUBDOMAIN="{{ cartodb_user_subdomain }}" PASSWORD="{{ cartodb_user_password }}" EMAIL="{{ cartodb_user_email }}"
```
Will this still work if the user already exists?

CloudNiner replied:
Yes, it happily runs the rake task but exits without doing anything

ddohler commented on ansible/roles/cartodb.postgresql/tasks/main.yml:
```diff
@@ -0,0 +1,55 @@
+ - name: Set postgresql.conf shared_preload_libraries
+   lineinfile: dest=/etc/postgresql/9.3/main/postgresql.conf regexp=^shared_preload_libraries insertafter=^#shared_preload_libraries line="shared_preload_libraries = 'schema_triggers.so'" state=present
+
+ - name: Restart postgresql
```
Looks like you've already got a handler to restart Postgres, so you can probably just `notify` it here (and in `postgis.yml` and `cartodb_extension.yml`.

hectcastro commented on README.md:
````diff
@@ -1,2 +1,35 @@
+
+Lastly, you'll need to add the following entries in your hosts file to point to the VM:
+```
+192.168.20.100 development.localhost.lan
````
If you need a hostname, and are willing to depend on xip.io, you can just use something like `192.168.20.100.xip.io`, which will resolve to `192.168.20.100` vs. requiring an `/etc/hosts` entry.

CloudNiner replied:
This might work, but would at the least require more configuration of cartodb. I'll tinker, but will likely not do this (although its nice to know about)

hectcastro commented on Vagrantfile:
```diff
@@ -0,0 +1,47 @@
+Vagrant.configure("2") do |config|
+
+  config.vm.define "cartodb" do |cartodb|
+    cartodb.vm.box = "precise64"
```
I think you can just use `ubuntu/precise64` here.

hectcastro commented on ansible/roles/cartodb.app.dependencies/tasks/main.yml:
```diff
@@ -0,0 +1,48 @@
+
+ # TODO: Replace with bundler module if we want to only use ansible 2.0.0+
+ - name: Install ruby bundle
+   shell: bash -lc "bundle install" chdir={{ cartodb_dir }}
```
Is this something to pick up the `/etc/environment` changes?

CloudNiner replied:
This is something that was in the other cartodb provisioning repo, just copied it from there. I'll see about cleaning this up.

hectcastro commented on ansible/roles/cartodb.app/templates/cartodb_server.conf.j2:
```diff
@@ -0,0 +1,17 @@
+script
+    APP_ROOT={{ cartodb_dir }}
+
+    cd $APP_ROOT
```
Might be better to use `chdir` here vs. depending on the ability to `cd` inside a `script` block.

CloudNiner replied:
Was able to swap to chdir both in cartodb_resque_worker and cartodb_server

hectcastro commented on ansible/roles/cartodb.app/templates/cartodb_server.conf.j2:
```diff
@@ -0,0 +1,17 @@
+    APP_ROOT={{ cartodb_dir }}
+
+    cd $APP_ROOT
+    sudo -u {{ ansible_ssh_user }} RAILS_ENV={{ cartodb_environment }} bundle exec rails server
```
Would `setuid` here work instead of `sudo -u`?

CloudNiner replied:
Just tried this, for some reason it doesn't pick up the locally installed bundle. Made the chdir change specified above, can investigate this later

hectcastro commented on ansible/roles/cartodb.postgresql/tasks/main.yml:
```diff
@@ -0,0 +1,55 @@
+     - Restart PostgreSQL
+
+ - name: Add cartodb publicuser
+   command: createuser publicuser --no-createrole --no-createdb --no-superuser
```
The `postgresql_user` Ansible module should be able to do this.

hectcastro commented on ansible/roles/cartodb.redis/tasks/main.yml:
```diff
@@ -0,0 +1,15 @@
+ #          By default redis server is configured to not have any type of disk persistence.
+ #          If stopped or restarted everything stored in redis will be lost.
+ #          In CartoDB redis is not just a simple cache storage. It stores information
+ #          that needs to be persisted.
```
:(

hectcastro commented:
Might make sense to rename the repository to `vagrant-cartodb` to address your concerns over it not aligning with the existing `ansible-*` repositories.

ddohler commented:
Looks good to me as a first start. I'll let this run overnight to test.

Commit:
PR feedback cleanup

- Use chdir directive in upstart jobs when able
- Revert to ansible command module for bundle install
- Use more generic "ubuntu/precise64" base box definition in Vagrantfile
- Update postgresql service restarts to use ansible notify
- Skip app_config.yml and database.yml copies if already exist
- Use template group_vars/all file
- Readd missing azavea.pip dependency
- Use file module to touch files instead of command
- Default ruby_development group_var to False
- Use ansible postgresql_user module to set tile/public users

CloudNiner commented:
I like the suggestion to do change the repo name to `vagrant-cartodb`, I'll do that once this PR is merged.

CloudNiner commented:
@hectcastro @ddohler Updated with a few more commits that should address almost all the comments and cleanup a few extra minor things. Feel free to take another look, I'm provisioning a fresh box atm.

CloudNiner commented:
Current changeset provisioned successfully, although I did get a pip error about mismatched hashes when downloading packages for the requirements.txt file. No issues on followup reprovision.

ddohler commented:
:+1:

Commit:
Update README instructions

Add git clone argument to checkout branch 3.12.2
Add note about updating sqlapi/windshaft/cdb-postgresql versions
  if the user changes the cartodb version

hectcastro commented on ansible/roles/cartodb.nodejs/tasks/main.yml:
```diff
@@ -0,0 +1,7 @@
+   apt_repository: repo=ppa:cartodb/nodejs-010 state=present update_cache=yes
+
+ - name: Install nodejs 0.10.x
+   apt: name=nodejs state=present
```
Is it worth pinning this to something just to prevent the wrong package from being installed? The CartoDB one will probably get installed every time as things stand now because of the version mismatch:

``` bash
vagrant@cartodb:~$ sudo apt-cache show nodejs | grep Version
Version: 0.10.26-1chl1~precise1
Version: 0.6.12~dfsg1-1ubuntu1
```

CloudNiner replied:
Hmmm the version concerns are valid. I'm not sure how this will shake out over time, although I expect cartodb to add new repos when they update versions, rather than updating the package in the repo, since the package version is in the repository name. I'm inclined to leave this as is for now and re-evaluate later if cartodb starts making significant changes to these repos later.

hectcastro commented on ansible/roles/cartodb.postgresql/tasks/main.yml:
```diff
@@ -0,0 +1,57 @@
+    - libpq-dev
+    - python-psycopg2
+    - postgresql-client-{{ postgresql_version }}
+    - postgresql-client-common
```
Some concern over package version number changes here:

``` bash
vagrant@cartodb:~$ sudo apt-cache show postgresql-client-common | grep Version
Version: 149.cdb12.4+3.93
Version: 129ubuntu1
Version: 129
```

hectcastro commented on ansible/roles/cartodb.postgresql/tasks/main.yml:
```diff
@@ -0,0 +1,57 @@
+ #   but they don't seem to trigger until at the very end of provisioning, which is too late
+ #   for this particular handler.
+ - name: Restart PostgreSQL
+   service: name=postgresql state=restarted
```
This is probably the best approach given you know exactly what needs to be restarted, but also though I would mention `flush_handlers` as a way to force execution of the notifications that have been queued up to that point:

``` yaml
- meta: flush_handlers
```

hectcastro commented on ansible/roles/cartodb.postgresql/tasks/postgis.yml:
```diff
@@ -0,0 +1,33 @@
+   when: createdb_template_postgis.changed
+
+ - name: Run ldconfig
+   command: ldconfig
```
I could also see this making sense as a handler that gets notified by the things that require it to be run.

CloudNiner replied:
Good point, although I'm going to leave as is, since it only needs to be run once and specifically after these steps.

hectcastro commented on ansible/roles/cartodb.resque/tasks/main.yml:
```diff
@@ -0,0 +1,15 @@
+---
+
+ - name: Ensure pid file dir exists
+   command: mkdir -p {{ cartodb_resque_pid_dir }}
```
Could also do:

``` yaml
- name: Create PID directory
  file: path="{{ cartodb_resque_pid_dir }}"
        state=directory
```

ddohler replied:
Just noticed that Ansible 2.0 has a nice warning about this:

```
TASK [cartodb.resque : Ensure pid file dir exists] *****************************
changed: [cartodb]
 [WARNING]: Consider using file module with state=directory rather than running
mkdir
```

CloudNiner replied:
Snazzy warning. Will definitely make this change, it slid under the radar with all the other things I was changing/copypasting

hectcastro commented on ansible/roles/cartodb.resque/tasks/main.yml:
```diff
@@ -0,0 +1,15 @@
+   template: src=cartodb_start_resque.conf.j2 dest=/etc/init/cartodb_start_resque.conf
+
+ - name: Restart resque workers
+   command: service cartodb_resque_worker restart WORKERNUM={{ item }}
```
Interesting approach here. I'm not sure I have a much better alternative to keep this in Ansible land.

hectcastro commented on ansible/roles/cartodb.resque/templates/cartodb_resque_worker.conf.j2:
```diff
@@ -0,0 +1,18 @@
+instance $WORKERNUM
+
+script
+    RESQUEPID1={{ cartodb_resque_pid_dir }}/resque$WORKERNUM.pid
```
This can be broken out into a top level stanza:

``` bash
env RESQUEPID1={{ cartodb_resque_pid_dir }}/resque$WORKERNUM.pid
```

Unsure there is much benefit in this case though. Might be more valuable if you need to add additional stanzas later (`pre-start`, `post-stop`, etc.).

hectcastro commented on ansible/roles/cartodb.sqlapi/tasks/main.yml:
```diff
@@ -0,0 +1,31 @@
+   sudo: False
+
+ - name: Touch CartoDB SQL API config file
+   command: touch config/environments/{{ cartodb_environment }}.js
```
Could also use the `file` module here too.

hectcastro commented on ansible/roles/cartodb.windshaft/tasks/main.yml:
```diff
@@ -0,0 +1,36 @@
+   sudo: False
+
+ - name: Touch CartoDB Windshaft config file
+   command: touch config/environments/{{ cartodb_environment }}.js
```
Same here about `file`.

hectcastro commented:
Provisioned clearly here too. Good job wrangling all of these things together. :+1:

CloudNiner commented:
Thanks for the second round of feedback, will make a few adjustments, then merge and update the repository name.
